README.md
### Hydra-Synth

Video synth engine for [hydra](https://github.com/ojack/hydra).

Currently experimental / in-progress.

This is the main logic of hydra packaged as a javascript module, intended for use within javascript projects. If you are looking to get started with hydra quickly, visit the [web editor](https://hydra.ojack.xyz) or the [main repo](https://github.com/ojack/hydra). To use hydra within atom, follow the instructions at https://github.com/ojack/hydra-examples.

![image of hydra in webpage](/assets/hydra-webpage.png?raw=true)

### To include in a webpage (bundled version):

Include the bundled version of this library in your html file:

```html
<script src="https://unpkg.com/hydra-synth"></script>
<script>
  // create a new hydra-synth instance
  var hydra = new Hydra({ detectAudio: false });
  osc(4, 0.1, 1.2).out();
</script>
```

You can see and remix a live example here: https://glitch.com/edit/#!/hydra-webpage

### To use as a module:

Download the module:

```
npm install --save hydra-synth
```

Include in your app:

```javascript
const Hydra = require('hydra-synth');

const hydra = new Hydra({ detectAudio: false });
osc(4, 0.1, 1.2).out();
```

The rest of this README is about configuring hydra-synth. For broader hydra documentation and usage, see [getting started](https://github.com/ojack/hydra#basic-functions), [interactive function documentation](https://ojack.xyz/hydra-functions/), and [Hydra Book (by Naoto Hieda)](https://hydra-book.naotohieda.com/#/).

#### API:

```javascript
const hydra = new Hydra([opts]);
```

create a new hydra instance

If `opts` is specified, the default options (shown below) will be overridden.

```javascript
{
  canvas: null, // canvas element to render to. If none is supplied, a canvas will be created and appended to the screen

  width: // defaults to canvas width when included, 1280 if not

  height: // defaults to canvas height when included, 720 if not

  autoLoop: true, // if true, will automatically loop using requestAnimationFrame.If set to false, you must implement your own loop function using the tick() method (below)

  makeGlobal: true, // if false, will not pollute global namespace (note: there are currently bugs with this)

  detectAudio: true, // recommend setting this to false to avoid asking for microphone

  numSources: 4, // number of source buffers to create initially

  numOutputs: 4, // number of output buffers to use. Note: untested with numbers other than 4. render() method might behave unpredictably

  extendTransforms: [] // An array of transforms to be added to the synth, or an object representing a single transform

  precision: null  // force precision of shaders, can be 'highp', 'mediump', or 'lowp' (recommended for ios). When no precision is specified, will use highp for ios, and mediump for everything else.

  pb = null, // instance of rtc-patch-bay to use for streaming
}

```

#### Custom render loop

You can use your own render loop for triggering hydra updates, instead of the automatic looping. To use, set autoLoop to false, and call

```javascript
hydra.tick(dt);
```

where dt is the time elapsed in milliseconds since the last update

### To develop:

```javascript
npm run dev
```

Sets up an example using hydra-synth that is automatically updated when source files are updated. It is possible to write test code by editing /example/index.js or by writing hydra code into the developer console.

#### Non-global mode [in progress]

If makeGlobal is set to false, buffers and functions can be accessed via the synth property of the hydra instance. Note that sources and buffers are contained in an array and accessed by index. E.g.:

```javascript
let synth = hydra.synth;
synth.osc().out();
synth.s0.initCam();
```


CHANGELOG.md
# Changelog

## [1.3.8] - 2021-04-12

### Changed

- uses absolute position for mouse
- updated documentation

## [1.3.7] - 2021-04-06

### Fixed

- bug in precision for shaders

### Changed

- when no precision is specified, uses "highp" on ios, and "mediump" on everything else

## [1.3.5] - 2020-11-06

- Fixed divide by 0 on smoothstep

## [1.3.2] - 2020-11-04

### Fixed

- Typo on rotate function
- update for setResolution()
- warning of undefined type

### Changed

- video default to muted

## [1.3.0] - 2020-06-10

### Changed

- wrapping for scroll, scrollX, modulateScroll functions

### Added

- initVideo(url)
- initImage(url)

## [1.1.8] - 2020-04-13

### Changed

- updated format for glsl-functions

## [1.1.7] - 2020-04-10

### Added

- added smooth(), ease(), and fit() to array utils

### Changed

- changed resize() to setResolution()

### Bugs / to do

- setResolution() not scaling correctly. check whether textures are being resized

## [1.1.6] - 2020-04-03

### Fixed

- error in eval-sandbox

## [1.1.5] - 2020-04-03

### Added

- fps to set target rendering speed `fps=30`, not setting a value or `fps=undefined` will render as fast as possible
- when canvas is stretched, uses pixelated rendering rather than blurry
- stats.fps shows current fps (read-only)
- update function called each time a new frame is rendered. can be used like

```
update = (dt) => {
     // something i want to do
}
```

### Fixed

- invalid function does not crash editor

## [1.1.4] - 2020-04-02

### Added

- hush() function clears all screens and stops cameras
- speed variable for controlling time

### Changed

- s0.initScreen() now possible with no extension installed (in chrome)
- source textures default to one pixel when no source is specified
- screen share working in FireFox
- s0.clear() stops webcam and clears texture
- removed dependency on webrtc-adapter

## [1.1.2] - 2020-04-02

### Changed

- changed format for defining custom functions

  2.0.0

- resize() function
- updates to MakeGlobal
- eval() function on hydra synth
- dynamically add functions
- need to require hydra.synth rather than just base repo
- width and height globally defined
- update function

to do:

- ShaderGenerator [WIP], runs serverside
- check for errors on eval
- add outputs (o0, etc) and sources to shader generator


dev/index.js
const Hydra = require('./../index.js');
const loop = require('raf-loop');

const HydraShaders = require('./../shader-generator.js');

function init() {
  ///  exampleNonGlobal()

  //  exampleExtendTransforms()
  // const shader = new HydraShaders()

  // let x = shader.eval('osc().out()')
  // console.log(x.frag, x.uniforms)
  //
  // let y = shader.eval(`
  //     let myFunc = () => 4
  //     osc(myFunc).out()
  // `)
  // console.log(y.frag, y.uniforms)
  //
  // let z = shader.eval(`
  //     src(s0).out()
  // `)
  // console.log(z.frag, z.uniforms)
  const canvas = document.createElement('canvas');
  canvas.style.backgroundColor = '#000';
  canvas.width = 800;
  canvas.height = 200;
  document.body.appendChild(canvas);

  // canvas.style.width = '100%'
  // canvas.style.height = '100%'
  //  exampleCustomCanvas()
  var hydra = new Hydra({ detectAudio: false, canvas: canvas });
  //  window.hydra = hydra

  shape()
    .scroll(
      () => mouse.x / width,
      () => mouse.y / height
    )
    .out();

  // gradient(0.4).out()
  //  voronoi().out()

  const debugLog = document.createElement('pre');
  const frag = noise().glsl()[0].frag;
  debugLog.innerText = frag;
  document.body.appendChild(debugLog);
  //exampleSmoothstep()

  // shape().modulateScrollX(osc()).out()

  //  exampleGetGLSL()
  // window.update = (dt) => draw.draw(time)

  // shape(3, f0(), 0.5).out()
  //
  //   exampleAddFunction(hydra)
  // //  exampleScreen()
  //
  // exampleCustomCanvas()
  // exampleSmoothing()
  ///var generator = new shader()

  // document.body.appendChild(canvas)
  //  var hydra = new Hydra({
  //  //  autoLoad: false
  //    canvas: canvas,
  //    enableStreamCapture: true,
  //    detectAudio: true,
  //    makeGlobal: false
  //  })
  //
  //  window.hydra = hydra
  //
  //  var sinN = v => (Math.sin(v)+1)/2
  //  var cosN = v => (Math.cos(v)+1)/2
  //

  //
  //  osc().out()
  //
  //  // set bpm
  //  //hydra.bpm(30)
  //
  //  var x = 0
  //  loop((dt) => {
  //    hydra.tick(dt)
  //  }).start()

  //osc(5).out()
}

function exampleVideo() {
  s0.initVideo('https://media.giphy.com/media/26ufplp8yheSKUE00/giphy.mp4');
  src(s0).out();
}

function exampleEasingFunctions() {
  //  //
  //  // // Example array sequences
  //  // shape([4, 5, 3]).out()
  //  //
  //  // // array easing
  //  // shape([4, 3, 2].ease('easeInQuad')).out()
  //  //
  //  // // array smoothing
  //  // shape([4, 3, 2].smooth()).out()
}

// fixing smoothstep issue so that 0 is not passed as a parameter
function exampleSmoothstep() {
  shape(4, 0.3, 0.01).out();
  shape(4, 0.5, 0).out();
  osc(89, 0.01, 1.8).luma(0.5, 0).out();
  osc(89, 0.01, 1.8).thresh(0.5, 0).out();
}

function exampleVideo() {
  s0.initVideo('https://media.giphy.com/media/AS9LIFttYzkc0/giphy.mp4');
  src(s0).out();
}

function exampleNonGlobal() {
  var hydra = new Hydra({ makeGlobal: false });
  hydra.synth.fps = 10;
  hydra.synth.osc().out();
  window.hydra = hydra;
}

function exampleExtendTransforms() {
  var hydra = new Hydra({
    extendTransforms: {
      name: 'myOsc', // name that will be used to access function as well as within glsl
      type: 'src', // can be src: vec4(vec2 _st), coord: vec2(vec2 _st), color: vec4(vec4 _c0), combine: vec4(vec4 _c0, vec4 _c1), combineCoord: vec2(vec2 _st, vec4 _c0)
      inputs: [
        {
          name: 'freq',
          type: 'float', // 'float'   //, 'texture', 'vec4'
          default: 0.2,
        },
        {
          name: 'sync',
          type: 'float',
          default: 0.1,
        },
        {
          name: 'offset',
          type: 'float',
          default: 0.0,
        },
      ],
      glsl: `
         vec2 st = _st;
        float r = sin((st.x-offset*20./freq-time*sync)*freq)*0.5  + 0.5;
        float g = sin((st.x+time*sync)*freq)*0.5 + 0.5;
        float b = sin((st.x+offset/freq+time*sync)*freq)*0.5  + 0.5;
        return vec4(r, g, b, 1.0);
       `,
    },
  });
  myOsc(10, 0.2, 0.8).out();
}

function exampleImage() {
  s0.initImage('https://upload.wikimedia.org/wikipedia/commons/2/25/Hydra-Foto.jpg');
  src(s0).out();
}

function exampleAddFunction(hydra) {
  // example custom function
  setFunction({
    name: 'myOsc', // name that will be used to access function as well as within glsl
    type: 'src', // can be src: vec4(vec2 _st), coord: vec2(vec2 _st), color: vec4(vec4 _c0), combine: vec4(vec4 _c0, vec4 _c1), combineCoord: vec2(vec2 _st, vec4 _c0)
    inputs: [
      {
        name: 'freq',
        type: 'float', // 'float'   //, 'texture', 'vec4'
        default: 0.2,
      },
      {
        name: 'sync',
        type: 'float',
        default: 0.1,
      },
      {
        name: 'offset',
        type: 'float',
        default: 0.0,
      },
    ],
    glsl: `
    vec2 st = _st;
   float r = sin((st.x-offset*20./freq-time*sync)*freq)*0.5  + 0.5;
   float g = sin((st.x+time*sync)*freq)*0.5 + 0.5;
   float b = sin((st.x+offset/freq+time*sync)*freq)*0.5  + 0.5;
   return vec4(r, g, b, 1.0);
  `,
  });

  myOsc(10, 0.2, 0.8).out();
  //
  //  // ooo(10, 0.01, 1.2).blur().out()
}

function exampleScreen() {
  s0.initScreen();
  //src(s0).out()
}

function exampleGetGLSL() {
  src(s0).blend(o0).glsl();
}

function exampleCustomCanvas() {
  const canvas = document.createElement('canvas');
  canvas.style.backgroundColor = '#000';
  canvas.width = 800;
  canvas.height = 200;

  canvas.style.width = '100%';
  canvas.style.height = '100%';

  //canvas.style.imageRe

  var ctx = canvas.getContext('2d');
  ctx.moveTo(0, 0);
  ctx.lineTo(200, 100);
  ctx.stroke();

  s0.init({ src: canvas });
}

function exampleSmoothing() {
  var shapes = [
    shape(4).scale(1, 0.5, [0.5, 1, 2]).scrollX(0.3),
    shape(4).scale(1, 0.5, [0.5, 1, 2].smooth(0.5)).scrollX(0.0),
    shape(4).scale(1, 0.5, [0.5, 1, 2].smooth()).scrollX(-0.3),
  ];

  solid().add(shapes[0]).add(shapes[1]).add(shapes[2]).out(o0);
}

function exampleSetResolution() {
  setResolution(20, 20);
}

window.onload = init;
